rust/src/lib.rs

```diff
@@ -385,6 +438,42 @@ pub unsafe extern "C" fn zcashlc_list_accounts(
     unwrap_exc_or_null(res)
 }
 
+/// Returns the account data for the specified account identifier, or `nullptr` if
+/// the account id does not correspond to an account in the wallet.
+///
+/// # Safety
+///
+/// - `db_data` must be non-null and valid for reads for `db_data_len` bytes, and it must have an
+///   alignment of `1`. Its contents must be a string representing a valid system path in the
+///   operating system's preferred representation.
+/// - The memory referenced by `db_data` must not be mutated for the duration of the function call.
+/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of pointer::offset.
+/// - `account_uuid_bytes` must be non-null and valid for reads for 16 bytes, and it must have an alignment
+///    of `1`.
+/// - Call [`zcashlc_free_account`] to free the memory associated with the returned pointer
+///   when done using it.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_get_account(
+    db_data: *const u8,
+    db_data_len: usize,
+    network_id: u32,
+    account_uuid_bytes: *const u8,
+) -> *mut FfiAccount {
+    let res = catch_panic(|| {
+        let network = parse_network(network_id)?;
+        let db_data = unsafe { wallet_db(db_data, db_data_len, network)? };
+        let account_uuid = account_uuid_from_bytes(account_uuid_bytes)?;
+
+        Ok(if let Some(account) = db_data.get_account(account_uuid)? {
+            Box::into_raw(Box::new(FfiAccount::from_account(&account)))
+        } else {
+            ptr::null_mut()
+        })
+    });
+    unwrap_exc_or_null(res)
```

Comment:
There's a representational problem here. You're returning `null` to mean both `Ok(None)` and `Err(_)`, which potentially means that an error can't be correctly detected on the Swift side.

In other APIs I've resolved this by adding `FfiFoobar::some` and `FfiFoobar::none` methods, with the latter generating a non-null `FfiFoobar` containing sentinel values indicating `None` (e.g. a ZIP 32 account index value that cannot be valid; e.g. cast `u32` to `i32` and then use the negative number space as the sentinel values). In other places, we have a natural wrapping struct that is always non-null that contains a nullable pointer to the optional struct, which also provides a distinguisher.

Reply:
Okay, I wasn't sure of the right thing to do here. I'll change this to a sentinel.